ext/odbc/php_odbc.c

```diff
@@ -49,23 +51,175 @@
 
 #include "odbc_arginfo.h"
 
-/*
- * not defined elsewhere
- */
+#define CHECK_ODBC_CONNECTION(conn) \
+	if (conn == NULL) { \
+		zend_throw_error(NULL, "ODBC connection has already been closed"); \
+		RETURN_THROWS(); \
+	}
 
-#ifndef TRUE
-#define TRUE 1
-#define FALSE 0
-#endif
+#define CHECK_ODBC_RESULT(result) \
+	if (result->conn_ptr == NULL) { \
+		zend_throw_error(NULL, "ODBC result has already been closed"); \
+		RETURN_THROWS(); \
+	}
 
 void odbc_do_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent);
+static void safe_odbc_disconnect(void *handle);
+static void close_results_with_connection(const odbc_connection *conn);
+
+static int le_pconn;
+
+static zend_class_entry *odbc_connection_ce, *odbc_result_ce;
+static zend_object_handlers odbc_connection_object_handlers, odbc_result_object_handlers;
+
+static void odbc_insert_new_result(zval *result) {
+	Z_TRY_ADDREF_P(result);
+	zend_hash_next_index_insert_new(&ODBCG(results), result);
+}
+
+static inline odbc_link *odbc_link_from_obj(zend_object *obj) {
+	return (odbc_link *)((char *)(obj) - XtOffsetOf(odbc_link, std));
+}
+
+#define Z_ODBC_LINK_P(zv) odbc_link_from_obj(Z_OBJ_P(zv))
+#define Z_ODBC_CONNECTION_P(zv) Z_ODBC_LINK_P(zv)->connection
+
+static zend_object *odbc_connection_create_object(zend_class_entry *class_type) {
+	odbc_link *intern = zend_object_alloc(sizeof(odbc_link), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_connection_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *odbc_connection_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Connection, use odbc_connect() or odbc_pconnect() instead");
+	return NULL;
+}
+
+static void odbc_link_free(odbc_link *link)
+{
+	ZEND_ASSERT(link->connection && "link has already been closed");
+
+	if (link->persistent) {
+		zend_hash_del(&EG(persistent_list), link->hash);
+	} else {
+		close_results_with_connection(link->connection);
+
+		zend_hash_del(&ODBCG(connections), link->hash);
+
+		/* If aborted via timer expiration, don't try to call any unixODBC function */
+		if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+			safe_odbc_disconnect(link->connection->hdbc);
+			SQLFreeConnect(link->connection->hdbc);
+			SQLFreeEnv(link->connection->henv);
+		}
+		efree(link->connection);
+		ODBCG(num_links)--;
+	}
+
+	link->connection = NULL;
+
+	zend_string_release(link->hash);
+	link->hash = NULL;
+}
+
+static void odbc_connection_free_obj(zend_object *obj)
+{
+	odbc_link *link = odbc_link_from_obj(obj);
+
+	if (!link->connection) {
+		zend_object_std_dtor(&link->std);
+		return;
+	}
+
+	odbc_link_free(link);
+	zend_object_std_dtor(&link->std);
```

Comment:
I'm fine going with your preference, it makes sense to me. P.S: I implemented this the current way because I adopted Nikita's style (https://github.com/php/php-src/commit/d96219c185e68c82beb994db2c93bd26f47ce16a), but I've just realized that I did the same way as you mentioned in case of pgsql: https://github.com/php/php-src/blob/81dc9359451a887ec3f025711e4c70e53a117114/ext/pgsql/pgsql.c#L202

Reply:
Early-return style is fine and I use it myself. In this particular case however, I found it adds more noise because both paths are quite simple, so by not using early-return style the  code becomes easier to read.